contracts/utils/introspection/ERC165Checker.sol

```diff
@@ -106,19 +109,34 @@ library ERC165Checker {
      * Interface identification is specified in ERC-165.
      */
     function supportsERC165InterfaceUnchecked(address account, bytes4 interfaceId) internal view returns (bool) {
-        // prepare call
-        bytes memory encodedParams = abi.encodeCall(IERC165.supportsInterface, (interfaceId));
+        (bool success, bool supported) = _trySupportsInterface(account, interfaceId);
+        return success && supported;
+    }
+
+    /**
+     * @dev Attempts to call `supportsInterface` on a contract and returns both the call
+     * success status and the interface support result.
+     *
+     * This function performs a low-level static call to the contract's `supportsInterface`
+     * function. It returns:
+     *
+     * * `success`: true if the call didn't revert, false if it did
+     * * `supported`: true if the call succeeded AND returned data indicating the interface is supported
+     */
+    function _trySupportsInterface(
+        address account,
+        bytes4 interfaceId
+    ) private view returns (bool success, bool supported) {
+        bytes4 selector = IERC165.supportsInterface.selector;
 
-        // perform static call
-        bool success;
-        uint256 returnSize;
-        uint256 returnValue;
         assembly ("memory-safe") {
-            success := staticcall(30000, account, add(encodedParams, 0x20), mload(encodedParams), 0x00, 0x20)
-            returnSize := returndatasize()
-            returnValue := mload(0x00)
+            mstore(0x00, selector)
+            mstore(0x04, interfaceId)
+            success := staticcall(30000, account, 0x00, 0x24, 0x00, 0x20)
+            supported := and(
+                gt(returndatasize(), 0x1F), // we have at least 20 bytes of returndata
+                iszero(iszero(mload(0x00))) // the first 20 bytes of returndata are non-zero
```

Comment:
why are we checking x!=0 as opposed to x==1?

Reply:
We want anything that's not 0 to be interpreted as true. I think the rationale was to avoid the `abi.decode` to revert (see [here](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3552)). With assembly that might not be relevant anymore.

Reply:
It's technically correct because the ERC only distinguishes between 0 and 1 (false and true), so it's unclear how to treat the entire 32-byte value returned. I agree it's not ideal.

I would keep it for backwards compatibilty

Reply:
It'll break one of our tests, though

https://github.com/OpenZeppelin/openzeppelin-contracts/pull/5880#discussion_r2297482485

I think we should weigh @Amxx opinion here